Makefile

```diff
@@ -1,47 +1,63 @@
-ifndef GOPATH
-$(error No GOPATH set)
-endif
+SHORT_NAME := helm
+DEIS_REGISTRY ?= ${DEV_REGISTRY}
+IMAGE_PREFIX ?= helm
+
+REPO_PATH := github.com/helm/${SHORT_NAME}
 
+# The following variables describe the containerized development environment
+# and other build options
 BIN_DIR := bin
 DIST_DIR := _dist
 GO_PACKAGES := action chart config dependency log manifest release plugins/sec plugins/example codec
 MAIN_GO := helm.go
 HELM_BIN := $(BIN_DIR)/helm
 PATH_WITH_HELM = PATH="$(shell pwd)/$(BIN_DIR):$(PATH)"
-
 VERSION := $(shell git describe --tags --abbrev=0 2>/dev/null)+$(shell git rev-parse --short HEAD)
+DEV_ENV_IMAGE := quay.io/deis/go-dev:0.9.0
+DEV_ENV_WORK_DIR := /go/src/${REPO_PATH}
+DEV_ENV_CMD := docker run --rm -v ${CURDIR}:${DEV_ENV_WORK_DIR} -w ${DEV_ENV_WORK_DIR} ${DEV_ENV_IMAGE}
+DEV_ENV_CMD_INT := docker run -it --rm -v ${CURDIR}:${DEV_ENV_WORK_DIR} -w ${DEV_ENV_WORK_DIR} ${DEV_ENV_IMAGE}
+LDFLAGS := "-s -X main.version=${VERSION}"
 
-export GO15VENDOREXPERIMENT=1
 
-ifndef VERSION
-  VERSION := git-$(shell git rev-parse --short HEAD)
-endif
 
-build: $(MAIN_GO)
-	go build -o $(HELM_BIN) -ldflags "-X github.com/helm/helm/cli.version=${VERSION}" $<
+# Allow developers to step into the containerized development environment
 
-bootstrap:
-	go get -u github.com/golang/lint/golint github.com/mitchellh/gox
-	glide install
+# Containerized dependency resolution
+
+bootstrap: check-docker 
+	${DEV_ENV_CMD} glide install
+
+check-docker:
+	@if [ -z $$(which docker) ]; then \
+		echo "Missing \`docker\` client which is required for development"; \
+		exit 2; \
+	fi
+# Containerized build of the binary
+build: check-docker
+	mkdir -p ${BIN_DIR}
+	make binary-build
+
+docker-build: build check-docker
+	docker build --rm -t ${IMAGE} rootfs
+	docker tag -f ${IMAGE} ${MUTABLE_IMAGE}
 
-build-all:
-	gox -verbose \
-	-ldflags "-X github.com/helm/helm/cli.version=${VERSION}" \
-	-os="linux darwin " \
-	-arch="amd64 386" \
-	-output="$(DIST_DIR)/{{.OS}}-{{.Arch}}/{{.Dir}}" .
 
-clean:
-	rm -rf $(DIST_DIR) $(BIN_DIR)
+# Builds the binary-- this should only be executed within the
+# containerized development environment.
 
-dist: build-all
```

Comment:
This target should not be removed.  And it's implementation is currently under `build`, instead of here where it belongs.  When that is fixed, there's also an opportunity to containerize the steps involving `zip`.